### PR TITLE
Cleanup empty namespaces

### PR DESCRIPTION
**Pull request by srueg**
## Summary

Cleanup empty namespaces.

## Checklist

- [x] Keep pull requests small so they can be easily reviewed.
- [x] Categorize the PR by setting a good title and adding one of the labels:
      `bug`, `enhancement`, `documentation`, `change`, `breaking`,
      as they show up in the changelog
- [x] Update the documentation.
- [x] Update tests.

### CODE REVIEW

File: pkg/namespace/namespace.go [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+			if now.After(deleteAt) {
+				emptyNamespaces = append(emptyNamespaces, ns)
+			}
+		} else {
+			nsCopy := ns.DeepCopy()
+			if nsCopy.Annotations == nil {
+				nsCopy.Annotations = make(map[string]string, 1)
+			}
+			nsCopy.Annotations[cleanAnnotation] = now.UTC().Format(util.TimeFormat)
+			log.Infof("Annotate namespace for deletion: %q", nsCopy.Name)
```
Comment (ccremer) [inline pkg/namespace/namespace.go:106]:
```suggestion
			log.Infof("Annotated namespace for deletion: %q", nsCopy.Name)
```

File: pkg/namespace/namespace_test.go [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+		"DeleteNotYetNamespaces": {
+			objs: []runtime.Object{
+				&corev1.Namespace{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: "test-ns-1",
+						Annotations: map[string]string{
+							cleanAnnotation: time.Now().UTC().Format(util.TimeFormat),
+						},
+					},
+				},
```
Comment (ccremer) [inline pkg/namespace/namespace_test.go:85]: I guess that could be put into a test helper function to remove clutter?

File: go.mod [hunk 3]
```diff
@@ -3,25 +3,26 @@ module github.com/appuio/seiso
 go 1.16
 
 require (
-	github.com/go-logr/logr v0.3.0 // indirect
-	github.com/googleapis/gnostic v0.5.1 // indirect
+	github.com/etdub/goparsetime v0.0.0-20160315173935-ea17b0ac3318 // indirect
 	github.com/hashicorp/go-version v1.3.0
-	github.com/imdario/mergo v0.3.10 // indirect
-	github.com/karrick/tparse v2.8.2+incompatible
+	github.com/karrick/tparse v2.4.2+incompatible
```
Comment (ccremer) [inline go.mod:8]: Why the downgrade?

File: cmd/namespace_test.go [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+		args   []string
+		config cfg.Configuration
+	}
+	tests := []struct {
+		name    string
+		input   args
+		wantErr bool
+	}{
+		{
+			name: "ShouldThrowError_IfNoLabelSelector",
```
Comment (ccremer) [inline cmd/namespace_test.go:23]: Could also follow the map[string]struct {} syntax.
```suggestion
	tests := map[string]struct {
		args []string
		config cfg.Configuration
		wantErr bool
	}{
		"ShouldThrowError_IfNoLabelSelector": {
```

**Review by ccremer — COMMENTED** [review]
Review is probably note-to-self :)

Might make sense to add integration tests or even e2e tests with KIND, seeing that helm testing fails.

File: cmd/namespaces.go [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+
+	"github.com/appuio/seiso/cfg"
+	"github.com/appuio/seiso/pkg/kubernetes"
+	"github.com/appuio/seiso/pkg/namespace"
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+)
+
+const (
+	nsCommandLongDescription = `Sometimes Namespcaes are left empty in a Kubernetes cluster.
```
Comment (cimnine) [inline cmd/namespaces.go:16]:
```suggestion
	nsCommandLongDescription = `Sometimes Namespaces are left empty in a Kubernetes cluster.
```

File: go.mod (hunk 3 above)
Comment (cimnine) [reply go.mod]: I guess it was due to missing new-style import, i.e. in `import github.com/karrick/tparse/v2`, the `/v2` part was missing.

**Review by simu — APPROVED** [review]
I didn't see any obvious problems in the code and the deletion logic looks correct.

File: cmd/namespaces.go [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+A Namespace is deemed empty if no Helm releases, Pods, Deployments, StatefulSets or DaemonSets can be found.`
+)
+
+var (
+	nsCmd = &cobra.Command{
+		Use:          "namespaces",
+		Short:        "Cleans up your empty Namespaces",
+		Long:         nsCommandLongDescription,
+		Aliases:      []string{"namespace", "ns"},
+		Args:         cobra.MaximumNArgs(1),
```
Comment (ccremer) [inline cmd/namespaces.go:27]: This doesn't seem to be effective nor used. If the label selector is missing, seiso exits with missing label error

File: pkg/namespace/namespace.go [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+		if err != nil {
+			return nil, fmt.Errorf("could not get %s resources %w", checker.Name(), err)
+		}
+	}
+
+	for _, ns := range namespaces {
+		if _, ok := nonEmptyNamespaces[ns.Name]; ok {
+			// Namespace is not empty
+			if _, ok := ns.Annotations[cleanAnnotation]; ok && !nss.configuration.Batch {
+				log.Warnf("Non empty namespace is annotated for deletion, skip: %q", ns.Name)
```
Comment (ccremer) [inline pkg/namespace/namespace.go:81]: The output is somewhat strange:
```console
INFO Seiso unknown, commit dirty, date today      
DEBU Using config                                  git="{0 . false version}" history="{3}" log="{debug false true}" namespace=default orphan="{1w ^[a-z0-9]{40}$}" resource="{[idunno=test] 1w 1s}"
DEBU Getting Namespaces                           
WARN Non empty namespace is annotated for deletion, skip: "seiso" 
INFO Showing results                               delete_after=1s
INFO Nothing found to be deleted.                 
```
I would reword it a bit:
```suggestion
				log.Warnf("Namespace is annotated for deletion, but not empty. Skipping %q", ns.Name)
```

File: pkg/namespace/namespace.go [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+}
+
+func (nss NamespacesService) Print(namespaces []corev1.Namespace) {
+	if len(namespaces) == 0 {
+		log.Info("Nothing found to be deleted.")
+	}
+
+	for _, ns := range namespaces {
+		if nss.configuration.Batch {
+			fmt.Println(ns.GetName())
```
Comment (ccremer) [inline pkg/namespace/namespace.go:137]: should be consistent with the usage above
```suggestion
			fmt.Println(ns.Name)
```

File: cmd/namespaces.go [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+		coreClient.Namespaces(),
+		dynamicClient,
+		namespace.ServiceConfiguration{
+			Batch: config.Log.Batch,
+		})
+
+	log.Debug("Getting Namespaces")
+	allNamespaces, err := service.List(ctx, toListOptions(c.Labels))
+	if err != nil {
+		return fmt.Errorf("could not retrieve Namespaces with labels %q: %w", c.Labels, err)
```
Comment (ccremer) [inline cmd/namespaces.go:84]: Either we use uppercase `Namespaces` everywhere or lowercase `namespaces`, but I would avoid mixing 2 different spellings for consistency reasons.
Can you check various references in this PR? I saw some of them used in other logs and errors.

**Review by ccremer — CHANGES_REQUESTED** [review]

File: cmd/namespaces.go (hunk 6 above)
Comment (cimnine) [reply cmd/namespaces.go:27]: It's like this in all the other definitions as well. I would therefore leave it in. (Or shall I remove it from all the other definitions as well, e.g. [1](https://github.com/appuio/seiso/blob/master/cmd/history.go#L21), [2](https://github.com/appuio/seiso/blob/master/cmd/configmaps.go#L26), [3](https://github.com/appuio/seiso/blob/master/cmd/common.go), …)

File: cmd/namespaces.go (hunk 9 above)
Comment (cimnine) [reply cmd/namespaces.go:84]: A quick search revealed that all but one occurences in a string of the term _namespace_ are spelled as _Namespace_.

File: cmd/namespaces.go (hunk 6 above)
Comment (ccremer) [reply cmd/namespaces.go:27]: Well, the history command requires an additional arg, see the usage documentation `history [NAMESPACE/IMAGE]`. The configmap and secrets do not seem to use it. So I would remove it where the args are not needed.

File: cmd/namespaces.go (hunk 9 above)
Comment (ccremer) [reply cmd/namespaces.go:84]: So I've just found this one by chance, great :P

**Review by ccremer — APPROVED** [review]
